### PR TITLE
Feat: Added Event Recorder to the Work Status Controller

### PR DESCRIPTION
**Pull request by aagusuab**
### Description of your changes

Fixes #
 - Added Event Recorder to the Work Status Controller
 - Added event creation for reconciliation success
I have:

- [x] Read and followed Caravel's [Code of conduct](https://github.com/Azure/k8s-work-api/blob/master/code-of-conduct.md).
- [x] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

File: tests/e2e/apply_test.go [hunk 1]
```diff
@@ -208,7 +208,7 @@ var (
 		Describe("deleted from the Hub", func() {
 			BeforeEach(func() {
 				// Todo - Replace with Eventually.
-				time.Sleep(2 * time.Second) // Give time for AppliedWork to be created.
+				time.Sleep(10 * time.Second) // Give time for AppliedWork to be created.
```
[the remark this answers is not in the document]
Comment (aagusuab) [reply tests/e2e/apply_test.go]: I agree, should be discussed later in the PR for e2e test.

File: pkg/controllers/work_status_controller.go [hunk 2]
```diff
@@ -59,6 +71,7 @@ func (r *WorkStatusReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	klog.InfoS("work status reconcile loop triggered", "item", req.NamespacedName)
 	work, appliedWork, err := r.fetchWorks(ctx, req.NamespacedName)
 	if err != nil {
+		r.recorder.Eventf(work, v1.EventTypeWarning, EventReasonFetchWorkFailed, "Fetching Work or Applied Work failed for %s", work.GetName())
```
[the remark this answers is not in the document]
Comment (aagusuab) [reply pkg/controllers/work_status_controller.go]: I have removed this event.

File: pkg/controllers/work_status_controller.go [hunk 3]
```diff
@@ -36,20 +38,28 @@ import (
 	workapi "sigs.k8s.io/work-api/pkg/apis/v1alpha1"
 )
 
+const (
+	EventReasonDeleteStaleWorkFailed    = "DeleteStaleWorkFailed"
```
Comment (Ealianis) [inline pkg/controllers/work_status_controller.go:42]: Where is the DeleteStaleWorkSuccess event?

File: pkg/controllers/work_status_controller.go [hunk 4]
```diff
@@ -70,6 +80,8 @@ func (r *WorkStatusReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	newRes, staleRes := r.calculateNewAppliedWork(work, appliedWork)
 	if err = r.deleteStaleWork(ctx, staleRes); err != nil {
 		klog.ErrorS(err, "failed to delete all the stale work", "work", req.NamespacedName)
+		r.recorder.Eventf(work, v1.EventTypeWarning, EventReasonDeleteStaleWorkFailed, "Deleting stale work %s failed", work.GetName())
```
Comment (Ealianis) [inline pkg/controllers/work_status_controller.go:83]: #### Comment > Style
Two different message styles:
- Deleting stale work _%s failed_
- Updating AppliedWork _failed for %s_

File: pkg/controllers/work_status_controller.go [hunk 5]
```diff
@@ -78,9 +90,11 @@ func (r *WorkStatusReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	appliedWork.Status.AppliedResources = newRes
 	if err = r.spokeClient.Status().Update(ctx, appliedWork, &client.UpdateOptions{}); err != nil {
 		klog.ErrorS(err, "update appliedWork status failed", "appliedWork", appliedWork.GetName())
+		r.recorder.Eventf(work, v1.EventTypeWarning, EventReasonUpdateAppliedWorkFailed, "Updating AppliedWork failed for %s", appliedWork.GetName())
 		return ctrl.Result{}, err
 	}
 
+	r.recorder.Eventf(work, v1.EventTypeNormal, EventReasonUpdateAppliedWorkSucceed, "Updating AppliedWork succeeded for %s", appliedWork.GetName())
```
Comment (Ealianis) [inline pkg/controllers/work_status_controller.go:97]: Comment > Grammar
The object has been updated. Use past-tense.

File: pkg/controllers/work_status_controller.go (hunk 4 above)
Comment (Ealianis) [reply pkg/controllers/work_status_controller.go:83]: Find a format that works:

Create     succesful  for   W
Retrieve  failed          for   X
Delete     succesful   for   Y
Update    failed          for   Z

Perhaps down the road we can utilize a helper library to standardize these and fully qualify the resources, for example:  Creation successful for  multicluster.x-k8s.io.v1alpha1.works/ namespace / name

File: pkg/controllers/work_status_controller.go (hunk 5 above)
Comment (aagusuab) [reply pkg/controllers/work_status_controller.go]: I have changed the events to be past-tense.

File: pkg/controllers/work_status_controller.go (hunk 4 above)
Comment (aagusuab) [reply pkg/controllers/work_status_controller.go]: Will keep in mind.

File: pkg/controllers/work_status_controller.go (hunk 3 above)
Comment (aagusuab) [reply pkg/controllers/work_status_controller.go]: I have included them to reflect the change in the work resource.

File: pkg/controllers/work_status_controller.go [hunk 6]
```diff
@@ -70,17 +80,23 @@ func (r *WorkStatusReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	newRes, staleRes := r.calculateNewAppliedWork(work, appliedWork)
 	if err = r.deleteStaleWork(ctx, staleRes); err != nil {
 		klog.ErrorS(err, "failed to delete all the stale work", "work", req.NamespacedName)
+		r.recorder.Eventf(work, v1.EventTypeWarning, EventReasonResourceUnsuccessfullyGarbageCollected, "Resource unsuccessfully garbage collected for Work %s", work.GetName())
+
 		// we can't proceed to update the applied
 		return ctrl.Result{}, err
+	} else if len(staleRes) > 0 && err == nil {
+		r.recorder.Eventf(work, v1.EventTypeNormal, EventReasonResourceSuccessfullyGarbageCollected, "Resource(s) successfully garbage collected for Work %s", work.GetName())
```
Comment (Ealianis) [inline pkg/controllers/work_status_controller.go:88]: #### Request > Include
Please add a //Todo that indicates that this event does not have the granularity of which manifests were garbage collected. There could a be subset M of N manifests. Ideally the delete stale work loop will record the event for each deleted resource.

**Review by Ealianis — APPROVED** [review]

File: pkg/controllers/work_status_controller.go (hunk 6 above)
Comment (aagusuab) [reply pkg/controllers/work_status_controller.go:89]: Added TODO

**Pull request merged**